components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,558 @@
+/// An attempt will be recorded first. If the attempted amount exceeds the available quotas of
+/// current epoch, the requester will register itself and sleep until next epoch.
+macro_rules! request_imp {
+    ($limiter:ident, $priority:ident, $amount:ident, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        let cached_bytes_per_refill =
+            $limiter.bytes_per_epoch[priority_idx].load(Ordering::Relaxed);
+        // rate limits are disabled
+        if cached_bytes_per_refill == 0 {
+            return $amount;
+        }
+        let amount = std::cmp::min($amount, cached_bytes_per_refill);
+        let bytes_through =
+            $limiter.bytes_through[priority_idx].fetch_add(amount, Ordering::AcqRel) + amount;
+        // we prefer not to partially return only a portion of requested bytes
+        if bytes_through <= cached_bytes_per_refill {
+            return amount;
+        }
+        let now = Instant::now_coarse();
+        let mut wait = Duration::from_millis(0);
+        // take a snapshot of pending bytes, which will be used to deduce when this request can
+        // be fully served.
+        let pending_snapshot = {
+            let mut locked = $limiter.protected.lock();
+            locked.pending_bytes[priority_idx] +=
+                std::cmp::min(bytes_through - cached_bytes_per_refill, amount);
+            let pending_snapshot = locked.pending_bytes[priority_idx];
+            if locked.next_refill_time <= now {
+                $limiter.refill(&mut locked, now);
+            } else {
+                wait += locked.next_refill_time - now;
+            }
+            pending_snapshot
+        };
+        // wait until our ticket can actually be served
+        wait += DEFAULT_REFILL_PERIOD * ((pending_snapshot - 1) / cached_bytes_per_refill) as u32;
+        do_sleep!(wait, $mode);
+        tls_collect_rate_limiter_request_wait($priority.as_str(), wait);
+        amount
+    }};
+}
+
+impl PriorityBasedIORateLimiter {
+    fn new() -> Self {
+        PriorityBasedIORateLimiter {
+            bytes_through: Default::default(),
+            bytes_per_epoch: Default::default(),
+            protected: Mutex::new(PriorityBasedIORateLimiterProtected::new()),
+        }
+    }
+
+    /// Dynamically changes the total IO flow threshold.
     #[allow(dead_code)]
-    pub fn reset(&self) {
-        for i in self.read.iter() {
-            i.store(0, Ordering::Relaxed);
+    fn set_bytes_per_sec(&self, bytes_per_sec: usize) {
+        let now = (bytes_per_sec as f64 * DEFAULT_REFILL_PERIOD.as_secs_f64()) as usize;
+        let before = self.bytes_per_epoch[IOPriority::High as usize].swap(now, Ordering::Relaxed);
+        if before == 0 || now == 0 {
+            // toggle on/off rate limit.
+            // we hold this lock so a concurrent refill can't negate our effort.
+            let _locked = self.protected.lock();
+            for p in &[IOPriority::Medium, IOPriority::Low] {
```

Comment:
why change medium and low only when zero

Reply:
Limits are propagated from high priority to low priority in `refill`. Even modified here, they will be overwritten later.